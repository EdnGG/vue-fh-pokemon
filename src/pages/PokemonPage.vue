<template>
  <h1 v-if="!pokemon">Loading...</h1>

  <div v-else>
    <h1>Hows this Pockemon?</h1>
    <!-- Todo: image -->
    <PokemonImage :pokemonId="pokemon.id" :showPokemon="showPokemon" />

    <!-- Todo optiones -->
    <!-- <PokemonOptions :pokemons="pokemonArr" @select-pokemon="checkAnswer($event)" /> -->
    <PokemonOptions :pokemons="pokemonArr" @select-pokemon="checkAnswer" />

    <template v-if="showAnswer">
      <h2 class="fade-in">{{ message }}</h2>
      <button @click="newGame">New Game</button>
    </template>
  </div>
</template>

<script>
import PokemonImage from "@/components/PokemonImage";
import PokemonOptions from "@/components/PokemonOptions";

import getpokemonOptions from "@/helpers/getPokemonOptions";

export default {
  name: "PokemonPage",
  components: {
    PokemonImage,
    PokemonOptions,
  },
  data() {
    return {
      pokemonArr: [],
      // pokemon que se necesita adivinar
      pokemon: null,
      showPokemon: false,
      showAnswer: false,
      message: "",
    };
  },
  methods: {
    newGame() {
      this.showAnswer = false;
      this.showPokemon = false;
      this.pokemon = null;
      this.pokemonArr = [];
      this.mixPokemonArray();
    },
    async mixPokemonArray() {
      this.pokemonArr = await getpokemonOptions();
      // Generar numero aleatorio entre el 0 y el 3
      const rndInt = Math.floor(Math.random() * 4);
      this.pokemon = this.pokemonArr[rndInt];
    },
    checkAnswer(id) {
      this.showPokemon = true;
      this.showAnswer = true;

      if (id === this.pokemon.id) {
        this.showPokemon = true;
        this.message = `Correct!, ${this.pokemon.name}`;
      } else {
        this.showPokemon = false;
        this.message = `Wrong, it was ${this.pokemon.name}`;
      }
    },
  },
  mounted() {
    this.mixPokemonArray();
  },
};
</script>
