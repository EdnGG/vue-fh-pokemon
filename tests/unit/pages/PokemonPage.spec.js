import { shallowMount, mount } from "@vue/test-utils";
import PokemonPage from "@/pages/PokemonPage.vue";
import { pokemons } from "../mocks/pokemons.mock";

describe("PokemonPage", () => {
  let wrapper;
  // let mixPokemonArraySpy;

  beforeEach(() => {
    // mixPokemonArraySpy = jest.spyOn(PokemonPage.methods, "mixPokemonArray");
    wrapper = shallowMount(PokemonPage);
  });

  test("must match with snapshot", () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  test("must call mixPokemomArray when the component is mounted", () => {
    const mixPokemonArraySpy = jest.spyOn(
      PokemonPage.methods,
      "mixPokemonArray"
    );
    // no se ocupa guardar el shadowmount en una variable
    const wrapper = shallowMount(PokemonPage);

    expect(mixPokemonArraySpy).toHaveBeenCalled();
  });

  test("must be match with snapshot when pokemons are loaded", () => {
    const wrapper = shallowMount(PokemonPage, {
      data() {
        return {
          pokemonaArr: pokemons,
          pokemon: pokemons[0],
          showPokemon: false,
          showAnswer: false,
          message: "",
        };
      },
    });
    expect(wrapper.html()).toMatchSnapshot();
  });

  test("must show components PokemonImage and PokemonOptions", () => {
    const wrapper = shallowMount(PokemonPage, {
      data() {
        return {
          pokemonaArr: pokemons,
          pokemon: pokemons[0],
          showPokemon: false,
          showAnswer: false,
          message: "",
        };
      },
    });

    expect(wrapper.find("pokemon-image-stub").exists()).toBeTruthy();
    expect(wrapper.find("pokemon-options-stub").exists()).toBeTruthy();

    expect(wrapper.find("pokemon-image-stub").attributes("pokemonid")).toBe(
      "5"
    );
    // expect(
    //   wrapper.find("pokemon-options-stub").attributes("pokemons")
    // ).toBeTruthy();  // Este test no pasa porque no coincide en el snapshot
  });

  test("pruebas con checkAnswer", async () => {
    const wrapper = shallowMount(PokemonPage, {
      data() {
        return {
          pokemonArr: pokemons,
          pokemon: pokemons[0],
          showPokemon: false,
          showAnswer: false,
          message: "",
        };
      },
    });

    await wrapper.vm.checkAnswer(5);

    expect(wrapper.find("h2").exists()).toBeTruthy();
    expect(wrapper.vm.showPokemon).toBe(true);
    expect(wrapper.find("h2").text()).toBe(`Correct!, ${pokemons[0].name}`);

    await wrapper.vm.checkAnswer(10);

    expect(wrapper.vm.message).toBe(`Wrong, it was ${pokemons[0].name}`);
  });
});
