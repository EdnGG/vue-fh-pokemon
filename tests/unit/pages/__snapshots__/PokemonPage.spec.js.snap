// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PokemonPage must be match with snapshot when pokemons are loaded 1`] = `
<div>
  <h1>Hows this Pockemon?</h1><!-- Todo: image -->
  <pokemon-image-stub pokemonid="5" showpokemon="false"></pokemon-image-stub><!-- Todo optiones -->
  <!-- <PokemonOptions :pokemons="pokemonArr" @select-pokemon="checkAnswer($event)" /> -->
  <pokemon-options-stub pokemons=""></pokemon-options-stub>
  <!--v-if-->
</div>
`;

exports[`PokemonPage must match with snapshot 1`] = `<h1>Loading...</h1>`;
